### PR TITLE
Add "kind" field to ModifierInvocation AST json element.

### LINKED ISSUE #11081: Constructor superclass invocations are mislabled as ModifierInvocation in AST
## Description

When a constructor explicitly invokes a superclass constructor, the AST output misrepresents the NodeType as `'ModifierInvocation'` rather than as a `'FunctionCall'` or, I suppose a new NodeType that should be specific to constructor invocations. 

## Environment

- Compiler version: v0.8.2
- Target EVM version (as per compiler settings): default
- Operating system: Linux

## Steps to Reproduce

Compiling the below code with `--ast-compact-json`:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

contract A {
    constructor(uint256 a) {}
}

contract B is A {
    constructor() A(1) {}
}
```

The NodeType that represents `A(1)` is a `ModifierInvocation`:

```
[
  {
    "arguments": [
      {
        "hexValue": "31",
        "id": 13,
        "isConstant": false,
        "isLValue": false,
        "isPure": true,
        "kind": "number",
        "lValueRequested": false,
        "nodeType": "Literal",
        "src": "142:1:0",
        "typeDescriptions": {
          "typeIdentifier": "t_rational_1_by_1",
          "typeString": "int_const 1"
        },
        "value": "1"
      }
    ],
    "id": 14,
    "modifierName": {
      "id": 12,
      "name": "A",
      "nodeType": "IdentifierPath",
      "referencedDeclaration": 8,
      "src": "140:1:0"
    },
    "nodeType": "ModifierInvocation",
    "src": "140:4:0"
  }
]
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/11081

### CODE REVIEW

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -453,10 +453,19 @@ bool ASTJsonConverter::visit(ModifierDefinition const& _node)
...
 		make_pair("arguments", _node.arguments() ? toJson(*_node.arguments()) : Json::nullValue)
-	});
+	};
+	Declaration const* declaration = _node.name().annotation().referencedDeclaration;
+	if (declaration)
+	{
+		if (dynamic_cast<ModifierDefinition const*>(declaration))
+			attributes.emplace_back("kind", "modifierInvocation");
+		else if (dynamic_cast<FunctionDefinition const*>(declaration))
+			attributes.emplace_back("kind", "baseConstructorSpecifier");
```
Comment (hrkrshnn) [inline libsolidity/ast/ASTJsonConverter.cpp]: Can we add an assert about `isConstructor()` here?

File: Changelog.md [hunk 2]
```diff
@@ -18,6 +18,8 @@ Bugfixes:
  * SMTChecker: Fix internal error on pushing to ``string`` casted to ``bytes``.
 
 AST Changes:
+ * ModifierInvocation: Add ``kind`` field which can be ``"modifierInvocation"`` or ``"baseConstructorSpecifier"``.
```
Comment (hrkrshnn) [inline Changelog.md]:
```suggestion
 * ModifierInvocation: Add ``kind`` field which can be ``modifierInvocation`` or ``baseConstructorSpecifier``.
```
Fine with leaving too.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 3]
```diff
@@ -453,10 +453,19 @@ bool ASTJsonConverter::visit(ModifierDefinition const& _node)
...
 bool ASTJsonConverter::visit(ModifierInvocation const& _node)
 {
-	setJsonNode(_node, "ModifierInvocation", {
+	std::vector<pair<string, Json::Value>> attributes{
 		make_pair("modifierName", toJson(_node.name())),
 		make_pair("arguments", _node.arguments() ? toJson(*_node.arguments()) : Json::nullValue)
-	});
+	};
+	Declaration const* declaration = _node.name().annotation().referencedDeclaration;
+	if (declaration)
```
Comment (ekpyron) [inline libsolidity/ast/ASTJsonConverter.cpp]:
```suggestion
	if (Declaration const* declaration = _node.name().annotation().referencedDeclaration)
```
?

**Review by ekpyron — DISMISSED** [review]
LGTM, although I'd agree with both comments by @hrkrshnn.

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp]: I'm not sure this works with "parseOnly".

Comment (chriseth) [reply libsolidity/ast/ASTJsonConverter.cpp]: I'll leave out "kind" if it is not a constructor.

**Comment by chriseth** [comment]
Updated.

**Review by hrkrshnn — APPROVED** [review]

**Pull request merged**